### PR TITLE
Hashing preprocessing layer to convert categorical inputs to zero-based integer values

### PR DESCRIPTION
**Pull request by workingloong**

### CODE REVIEW

File: elasticdl_preprocessing/layers/hashing.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        # Converts integer inputs to string.
+        if inputs.dtype.is_integer:
+            if isinstance(inputs, tf.SparseTensor):
+                inputs = tf.SparseTensor(
+                    indices=inputs.indices,
+                    values=tf.as_string(inputs.values),
+                    dense_shape=inputs.dense_shape,
+                )
+            else:
+                inputs = tf.as_string(inputs)
```
Comment (brightcoder01) [inline elasticdl_preprocessing/layers/hashing.py:58]: Can tf.as_string handle RaggedTensor？

File: elasticdl_preprocessing/tests/hashing_test.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        hash_out = hash_layer(inp)
+        expected_out = np.array([[1], [0], [1], [1], [2]])
+        self.assertTrue(np.array_equal(hash_out.numpy(), expected_out))
+
+        ragged_in = tf.ragged.constant([["A", "B"], ["C", "D"], ["E"], []])
+        hash_out = hash_layer(ragged_in)
+        expected_ragged_out = tf.ragged.constant(
+            [[1, 0], [1, 1], [2], []], dtype=tf.int64
+        )
+        self.assertTrue(ragged_tensor_equal(hash_out, expected_ragged_out))
```
Comment (brightcoder01) [inline elasticdl_preprocessing/tests/hashing_test.py:26]: Check hash_out to be RaggedTensor?

File: elasticdl_preprocessing/tests/hashing_test.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+        sparse_in = ragged_in.to_sparse()
+        hash_out = hash_layer(sparse_in)
+        expected_sparse_out = expected_ragged_out.to_sparse()
+        self.assertTrue(sparse_tensor_equal(hash_out, expected_sparse_out))
```
Comment (brightcoder01) [inline elasticdl_preprocessing/tests/hashing_test.py:31]: Check hash_out to be SparseTensor?

File: elasticdl_preprocessing/layers/hashing.py (hunk 1 above)
Comment (workingloong) [reply elasticdl_preprocessing/layers/hashing.py:58]: Yes, the example:
```python
import tensorflow as tf
rag_tensor = tf.ragged.constant([[1.9],[4.6,7.9]])
str_rag_tensor = tf.as_string(rag_tensor)
print(str_rag_tensor)
print(str_rag_tensor.dtype)
```

File: elasticdl_preprocessing/tests/hashing_test.py (hunk 2 above)
Comment (workingloong) [reply elasticdl_preprocessing/tests/hashing_test.py:26]: `ragged_tensor_equal` will check it.
The codes in #1846 
https://github.com/sql-machine-learning/elasticdl/blob/e580f27bb95ab7b998ec436a4f688fafcea2d7c6/elasticdl_preprocessing/tests/test_utils.py#L22-L25

File: elasticdl_preprocessing/tests/hashing_test.py (hunk 3 above)
Comment (workingloong) [reply elasticdl_preprocessing/tests/hashing_test.py:31]: Yes, `sparse_tensor_equal` will check it. See the codes in #1846
https://github.com/sql-machine-learning/elasticdl/blob/e580f27bb95ab7b998ec436a4f688fafcea2d7c6/elasticdl_preprocessing/tests/test_utils.py#L5-L8

File: elasticdl_preprocessing/layers/hashing.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    The output will be `[[1], [0], [1], [1], [2]]`
+
+    Arguments:
+        num_bins: Number of hash bins.
+        **kwargs: Keyword arguments to construct a layer.
+
+    Input: A string, int32 or int64 `tf.Tensor`,
+        `tf.SparseTensor` or `tf.RaggedTensor`
+
+    Output: An int64 tensor with the same type as input.
```
Comment (brightcoder01) [inline elasticdl_preprocessing/layers/hashing.py]: An int64 tensor with the same type as input
=>
An int64 tensor with the same shape as input

**Review by brightcoder01 — APPROVED** [review]

**Pull request merged**